core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -321,6 +321,12 @@ class BlockManagerMasterEndpoint(
   }
 
   private def removeShuffle(shuffleId: Int): Future[Seq[Boolean]] = {
+    val mergerLocations =
+      if (Utils.isPushBasedShuffleEnabled(conf, isDriver)) {
+        mapOutputTracker.getShufflePushMergerLocations(shuffleId)
+      } else {
+        Seq.empty[BlockManagerId]
+      }
```

Comment:
For `def removeShuffle(shuffleId: Int): Future[Seq[Boolean]]`  in `BlockManagerMasterEndpoint`, I prefer to put this code before 
```
    val removeMsg = RemoveShuffle(shuffleId)
    val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
      bm.storageEndpoint.ask[Boolean](removeMsg).recover {
        // use false as default value means no shuffle data were removed
        handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
      }
    }.toSeq
```
because mapOutputTracker may unregisterShuffle this shuffle in `BlockManagerStorageEndpoint` before `mapOutputTracker.getShufflePushMergerLocations(shuffleId)`.
```
    case RemoveShuffle(shuffleId) =>
      doAsync[Boolean]("removing shuffle " + shuffleId, context) {
        if (mapOutputTracker != null) {
          mapOutputTracker.unregisterShuffle(shuffleId)
        }
        SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
      }
```

Reply:
IIUC below sends the message to all the executors for removing shuffle
```
  val removeMsg = RemoveShuffle(shuffleId)
    val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
      bm.storageEndpoint.ask[Boolean](removeMsg).recover {
        // use false as default value means no shuffle data were removed
        handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
      }
    }.toSeq
```



In the  `BlockManagerStorageEndpoint`, the below will remove the statuses cached in the executors and not the driver.
 
      case RemoveShuffle(shuffleId) =>
      doAsync[Boolean]("removing shuffle " + shuffleId, context) {
        if (mapOutputTracker != null) {
          mapOutputTracker.unregisterShuffle(shuffleId)
        }
        SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
      }
      
So, I don't think this should impact the implementation of `removeShuffle` in the `BlockManagerMasterEndpoint`. 
 cc. @mridulm

Reply:
I agree with @otterc, I dont see an issue with interleaving modifications impacting this - so we should be able to move it closer to where it is used.
Keeping it closer is preferable - though I would call it a nit, like how Chandni has characterized it.
We can also pull it within the `externalBlockStoreClient.map {` to tighten the visibility.

Reply:
The following is my debug logs:
```
23/01/11 21:34:33 WARN [block-manager-storage-async-thread-pool-120] storage.BlockManagerStorageEndpoint:72 : Call mapOutputTracker.unregisterShuffle for 12
23/01/11 21:34:33 WARN [dispatcher-BlockManagerMaster] storage.BlockManagerMasterEndpoint:72 : Call getShufflePushMergerLocations for 12
23/01/11 21:34:33 WARN [block-manager-storage-async-thread-pool-74] storage.BlockManagerStorageEndpoint:72 : Call mapOutputTracker.unregisterShuffle for 11
23/01/11 21:34:33 WARN [dispatcher-BlockManagerMaster] storage.BlockManagerMasterEndpoint:72 : Call getShufflePushMergerLocations for 11
```

If we move the code after RemoveShuffle RPC to BlockManagerStorageEndpoint, we may get empty merge locations.
```
    val removeMsg = RemoveShuffle(shuffleId)
    val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
      bm.storageEndpoint.ask[Boolean](removeMsg).recover {
        // use false as default value means no shuffle data were removed
        handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
      }
    }.toSeq
```

Reply:
I am not sure I understood how we are getting empty mergers, can you please give details @wankunde ? Thx

Reply:
To give context, `mapOutputTrackerMaster.unregisterShuffle` happens after `shuffleDriverComponents.removeShuffle` - and BlockManagerMaster.removeShuffle does an inf wait for `RemoveShuffle` to complete.
So we should have non-empty mergers, since `shuffleStatuses` has not yet been cleaned up for the shuffle id.

The `RemoveShuffle` from within `master.removeShuffle` sends `RemoveShuffle` to workers - where it is simply cleanup the client side maps.

Let me know if I am missing something !

Reply:
1.  BlockManagerMaster is also registered as a BlockManagerInfo
```
SparkContext: _env.blockManager.initialize(_applicationId)
BlockManager:

    val idFromMaster = master.registerBlockManager(
      id,
      diskBlockManager.localDirsString,
      maxOnHeapMemory,
      maxOffHeapMemory,
      storageEndpoint)
```

2. BlockManagerMaster receive RemoveShuffle RPC and remove all the shuffle data in MapOutputTrackerMaster
```
BlockManagerStorageEndpoint:
    case RemoveShuffle(shuffleId) =>
      doAsync[Boolean]("removing shuffle " + shuffleId, context) {
        if (mapOutputTracker != null) {
          logWarning(s"Call mapOutputTracker.unregisterShuffle for $shuffleId")
          mapOutputTracker.unregisterShuffle(shuffleId)
        }
-----------
MapOutputTrackerMaster:
  def unregisterShuffle(shuffleId: Int): Unit = {
    shuffleStatuses.remove(shuffleId).foreach { shuffleStatus =>
....
```
3.  No shuffle status in MapOutputTracker:
```
BlockManagerMasterEndpoint:

    val mergerLocations =
      if (Utils.isPushBasedShuffleEnabled(conf, isDriver)) {
        mapOutputTracker.getShufflePushMergerLocations(shuffleId)
      } else {
        Seq.empty[BlockManagerId]
      }

---------------
MapOutputTrackerMaster:

  override def getShufflePushMergerLocations(shuffleId: Int): Seq[BlockManagerId] = {
    logWarning("shuffleStatuses.get(shuffleId).map(_.getShufflePushMergerLocations) = " +
      s"${shuffleStatuses.get(shuffleId).map(_.getShufflePushMergerLocations)}")
    shuffleStatuses.get(shuffleId).map(_.getShufflePushMergerLocations).getOrElse(Seq.empty)
  }
```

4. It seems removeShuffleFromShuffleServicesFutures will also be empty due to this reason.

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L336

Reply:
`RemoveShuffle` is sent to the storage endpoint - not to the block manager.

You can test it with this diff:

```
diff --git a/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala b/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
index 9f03228bb4f..e04ef7f11db 100644
--- a/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
+++ b/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala
@@ -321,12 +321,6 @@ class BlockManagerMasterEndpoint(
   }
 
   private def removeShuffle(shuffleId: Int): Future[Seq[Boolean]] = {
-    val mergerLocations =
-      if (Utils.isPushBasedShuffleEnabled(conf, isDriver)) {
-        mapOutputTracker.getShufflePushMergerLocations(shuffleId)
-      } else {
-        Seq.empty[BlockManagerId]
-      }
     val removeMsg = RemoveShuffle(shuffleId)
     val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
       bm.storageEndpoint.ask[Boolean](removeMsg).recover {
@@ -374,6 +368,13 @@ class BlockManagerMasterEndpoint(
 
     val removeShuffleMergeFromShuffleServicesFutures =
       externalBlockStoreClient.map { shuffleClient =>
+        val mergerLocations = {
+          if (Utils.isPushBasedShuffleEnabled(conf, isDriver)) {
+            mapOutputTracker.getShufflePushMergerLocations(shuffleId)
+          } else {
+            Seq.empty[BlockManagerId]
+          }
+        }
         mergerLocations.map { bmId =>
           Future[Boolean] {
             shuffleClient.removeShuffleMerge(bmId.host, bmId.port, shuffleId,
diff --git a/core/src/test/resources/log4j2.properties b/core/src/test/resources/log4j2.properties
index ab02104c696..b906523789d 100644
--- a/core/src/test/resources/log4j2.properties
+++ b/core/src/test/resources/log4j2.properties
@@ -16,7 +16,7 @@
 #
 
 # Set everything to be logged to the file target/unit-tests.log
-rootLogger.level = info
+rootLogger.level = debug
 rootLogger.appenderRef.file.ref = ${sys:test.appender:-File}
 
 appender.file.type = File
diff --git a/core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala b/core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala
index a13527f4b74..fb9dc8ff29b 100644
--- a/core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala
+++ b/core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala
@@ -17,9 +17,12 @@
 
 package org.apache.spark
 
+import java.util.{Collections => JCollections, HashSet => JHashSet}
 import java.util.concurrent.atomic.LongAdder
 
+import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
+import scala.collection.concurrent
 
 import org.mockito.ArgumentMatchers.any
 import org.mockito.Mockito._
@@ -30,10 +33,12 @@ import org.apache.spark.broadcast.BroadcastManager
 import org.apache.spark.internal.config._
 import org.apache.spark.internal.config.Network.{RPC_ASK_TIMEOUT, RPC_MESSAGE_MAX_SIZE}
 import org.apache.spark.internal.config.Tests.IS_TESTING
+import org.apache.spark.network.shuffle.ExternalBlockStoreClient
 import org.apache.spark.rpc.{RpcAddress, RpcCallContext, RpcEnv}
 import org.apache.spark.scheduler.{CompressedMapStatus, HighlyCompressedMapStatus, MapStatus, MergeStatus}
 import org.apache.spark.shuffle.FetchFailedException
-import org.apache.spark.storage.{BlockManagerId, ShuffleBlockId, ShuffleMergedBlockId}
+import org.apache.spark.storage.{BlockManagerId, BlockManagerInfo, BlockManagerMaster, BlockManagerMasterEndpoint, ShuffleBlockId, ShuffleMergedBlockId}
+import org.mockito.invocation.InvocationOnMock
 
 class MapOutputTrackerSuite extends SparkFunSuite with LocalSparkContext {
   private val conf = new SparkConf
@@ -913,9 +918,64 @@ class MapOutputTrackerSuite extends SparkFunSuite with LocalSparkContext {
     slaveRpcEnv.shutdown()
   }
 
+  test("SPARK-40480: shuffle remove should cleanup merged files as well") {
+
+    val newConf = new SparkConf
+    newConf.set("spark.shuffle.push.enabled", "true")
+    newConf.set("spark.shuffle.service.enabled", "true")
+    newConf.set(SERIALIZER, "org.apache.spark.serializer.KryoSerializer")
+    newConf.set(IS_TESTING, true)
+
+    val SHUFFLE_ID = 10
+
+    // needs TorrentBroadcast so need a SparkContext
+    withSpark(new SparkContext("local", "MapOutputTrackerSuite", newConf)) { sc =>
+      val rpcEnv = sc.env.rpcEnv
+      val masterTracker = sc.env.mapOutputTracker.asInstanceOf[MapOutputTrackerMaster]
+
+      val blockStoreClient = mock(classOf[ExternalBlockStoreClient])
+      val blockManagerMasterEndpoint = new BlockManagerMasterEndpoint(
+        rpcEnv,
+        sc.isLocal,
+        sc.conf,
+        sc.listenerBus,
+        Some(blockStoreClient),
+        // We dont care about this ...
+        new concurrent.TrieMap[BlockManagerId, BlockManagerInfo](),
+        masterTracker,
+        sc.env.shuffleManager,
+        true
+      )
+      rpcEnv.stop(sc.env.blockManager.master.driverEndpoint)
+      sc.env.blockManager.master.driverEndpoint =
+        rpcEnv.setupEndpoint(BlockManagerMaster.DRIVER_ENDPOINT_NAME,
+          blockManagerMasterEndpoint)
+
+      masterTracker.registerShuffle(SHUFFLE_ID, 10, 10)
+      val mergerLocs = (1 to 10).map(x => BlockManagerId(s"exec-$x", s"host-$x", x))
+      masterTracker.registerShufflePushMergerLocations(SHUFFLE_ID, mergerLocs)
+
+      assert(masterTracker.getShufflePushMergerLocations(SHUFFLE_ID).map(_.host).toSet ==
+        mergerLocs.map(_.host).toSet)
+
+      val foundHosts = JCollections.synchronizedSet(new JHashSet[String]())
+      when(blockStoreClient.removeShuffleMerge(any(), any(), any(), any())).thenAnswer(
+        (m: InvocationOnMock) => {
+          val host = m.getArgument(0).asInstanceOf[String]
+          val shuffleId = m.getArgument(2).asInstanceOf[Int]
+          assert(shuffleId == SHUFFLE_ID)
+          foundHosts.add(host)
+          true
+      })
+
+      sc.cleaner.get.doCleanupShuffle(SHUFFLE_ID, blocking = true)
+      assert(foundHosts.asScala == mergerLocs.map(_.host).toSet)
+    }
+  }
+
   test("SPARK-34826: Adaptive shuffle mergers") {
     val newConf = new SparkConf
-    newConf.set("spark.shuffle.push.based.enabled", "true")
+    newConf.set("spark.shuffle.push.enabled", "true")
     newConf.set("spark.shuffle.service.enabled", "true")
 
     // needs TorrentBroadcast so need a SparkContext




```

Reply:
Thinking more, I think your assessment is right : will need to check more.
If right, we need to the first `bm.storageEndpoint` to the end of the method.

The test above is slightly broken - the initial initialization is not getting handled (when driver would be registered).
A bit late for me to try to fix it :-)

Can you take a look at it @wankunde ? We can fix the test and add to the PR ?

Reply:
Thanks @mridulm , add this UT.
Yes, `RemoveShuffle` is sent to the storage endpoint in `removeShuffle`.
Add more debug log in our test cluster, shuffle 2 will be removed by thread `block-manager-storage-async-thread-pool-51` in `BlockManagerStorageEndpoint` class
```
23/01/12 00:43:40 WARN [block-manager-storage-async-thread-pool-51] storage.BlockManagerStorageEndpoint:72 : Call mapOutputTracker.unregisterShuffle for 2
23/01/12 00:43:40 WARN [block-manager-storage-async-thread-pool-51] spark.MapOutputTrackerMaster:72 : shuffleStatuses.remove(shuffleId) for shuffle 2
23/01/12 00:43:40 WARN [dispatcher-BlockManagerMaster] storage.BlockManagerMasterEndpoint:72 : Call getShufflePushMergerLocations for 2
23/01/12 00:43:40 WARN [dispatcher-BlockManagerMaster] spark.MapOutputTrackerMaster:72 : shuffleStatuses.get(shuffleId) = None
23/01/12 00:43:40 WARN [dispatcher-BlockManagerMaster] spark.MapOutputTrackerMaster:72 : shuffleStatuses.get(shuffleId).map(_.getShufflePushMergerLocations) = None
23/01/12 00:43:40 WARN [Spark Context Cleaner] spark.MapOutputTrackerMaster:72 : shuffleStatuses.remove(shuffleId) for shuffle 2
```